code/datums/elements/footstep.dm

```diff
@@ -0,0 +1,153 @@
+	var/footstep_type
+	///This can be a list OR a soundfile OR null. Determines whatever sound gets played.
+	var/footstep_sounds
+	///Whether or not to add variation to the sounds played
+	var/sound_vary = FALSE
+
+/datum/element/footstep/Attach(datum/target, footstep_type = FOOTSTEP_MOB_BAREFOOT, volume = 0.5, e_range = -8, sound_vary = FALSE)
+	. = ..()
+	if(!ismovable(target))
+		return ELEMENT_INCOMPATIBLE
+	src.volume = volume
+	src.e_range = e_range
+	src.footstep_type = footstep_type
+	src.sound_vary = sound_vary
+	switch(footstep_type)
+		if(FOOTSTEP_MOB_HUMAN)
+			if(!ishuman(target))
+				return ELEMENT_INCOMPATIBLE
+			RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/play_humanstep)
+			steps_for_living[target] = 0
+			return
+		if(FOOTSTEP_MOB_CLAW)
+			footstep_sounds = GLOB.clawfootstep
+		if(FOOTSTEP_MOB_BAREFOOT)
+			footstep_sounds = GLOB.barefootstep
+		if(FOOTSTEP_MOB_HEAVY)
+			footstep_sounds = GLOB.heavyfootstep
+		if(FOOTSTEP_MOB_SHOE)
+			footstep_sounds = GLOB.footstep
+		if(FOOTSTEP_MOB_SLIME)
+			footstep_sounds = 'sound/effects/footstep/slime1.ogg'
+		if(FOOTSTEP_OBJ_MACHINE)
+			footstep_sounds = 'sound/effects/bang.ogg'
+			RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/play_simplestep_machine)
+			return
+		if(FOOTSTEP_OBJ_ROBOT)
+			footstep_sounds = 'sound/effects/tank_treads.ogg'
+			RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/play_simplestep_machine)
+			return
+	RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/play_simplestep)
+	steps_for_living[target] = 0
+
+/datum/element/footstep/Detach(atom/movable/source)
+	UnregisterSignal(source, COMSIG_MOVABLE_MOVED)
+	steps_for_living -= source
+	return ..()
+
+///Prepares a footstep for living mobs. Determines if it should get played. Returns the turf it should get played on. Note that it is always a /turf/open
+/datum/element/footstep/proc/prepare_step(mob/living/source)
+	var/turf/open/turf = get_turf(source)
+	if(!istype(turf))
+		return
+
+	if(!turf.footstep || source.buckled || source.throwing || source.movement_type & (VENTCRAWLING | FLYING) || HAS_TRAIT(source, TRAIT_IMMOBILIZED))
+		return
+
+	if(source.body_position == LYING_DOWN) //play crawling sound if we're lying
+		playsound(turf, 'sound/effects/footstep/crawl1.ogg', 15 * volume, falloff_distance = 1, vary = sound_vary)
+		return
+
+	if(iscarbon(source))
+		var/mob/living/carbon/carbon_source = source
+		if(!carbon_source.get_bodypart(BODY_ZONE_L_LEG) && !carbon_source.get_bodypart(BODY_ZONE_R_LEG))
+			return
+		if(carbon_source.m_intent == MOVE_INTENT_WALK)
+			return// stealth
+	var/steps = ++steps_for_living[source]
+
+	if(steps >= 6)
+		steps = steps_for_living[source] = 0
+
+	if(steps % 2)
+		return
+
+	if(steps != 0 && !source.has_gravity(turf)) // don't need to step as often when you hop around
+		return
+	return turf
+
+/datum/element/footstep/proc/play_simplestep(mob/living/source)
+	SIGNAL_HANDLER
+
+	var/turf/open/turf = prepare_step(source)
```

Comment:
Needs a better name. *What* turf does this relate to? What they just stepped on? Where they were before? etc.